File: frontend/src/utils/recurrenceExpander.js
/**
 * Utility to expand recurring events into individual occurrences for calendar display.
 * Parses RFC 5545 RRULE format and generates event instances for a given date range.
 */

/**
 * Expands a recurring event into individual occurrences for the visible date range.
 * @param {Object} event - Event object with startTime, endTime, and recurrenceRule
 * @param {Date} rangeStart - Start of the visible date range
 * @param {Date} rangeEnd - End of the visible date range
 * @returns {Array} Array of expanded event instances
 */
export function expandRecurringEvent(event, rangeStart, rangeEnd) {
  if (!event.recurrenceRule || !event.startTime) {
    // Not a recurring event, return as single event
    return [event]
  }

  const occurrences = []
  const startDate = new Date(event.startTime)
  const endDate = new Date(event.endTime)
  const duration = endDate - startDate // Duration in milliseconds

  // Parse RRULE
  const rrule = parseRRULE(event.recurrenceRule)
  
  if (!rrule) {
    // Invalid RRULE, return original event
    return [event]
  }

  // Generate occurrences based on frequency
  let currentDate = new Date(startDate)
  const maxOccurrences = 100 // Limit to prevent infinite loops
  let count = 0

  while (currentDate <= rangeEnd && count < maxOccurrences) {
    // Check if this occurrence is within the visible range
    if (currentDate >= rangeStart) {
      const occurrence = {
        ...event,
        id: `${event.id}-${currentDate.getTime()}`, // Unique ID for this occurrence
        startTime: new Date(currentDate),
        endTime: new Date(currentDate.getTime() + duration),
        isRecurringInstance: true,
        originalEventId: event.id
      }
      occurrences.push(occurrence)
    }

    // Calculate next occurrence based on frequency
    currentDate = getNextOccurrence(currentDate, rrule, startDate)
    count++
  }

  return occurrences.length > 0 ? occurrences : [event]
}

/**
 * Parses RFC 5545 RRULE format into an object.
 * @param {string} rruleString - RRULE string (e.g., "FREQ=WEEKLY;BYDAY=MO")
 * @returns {Object} Parsed RRULE object
 */
function parseRRULE(rruleString) {
  if (!rruleString) return null

  const rrule = {}
  const parts = rruleString.split(';')

  for (const part of parts) {
    const [key, value] = part.split('=')
    if (key && value) {
      rrule[key.toUpperCase()] = value.toUpperCase()
    }
  }

  return rrule
}

/**
 * Calculates the next occurrence date based on RRULE.
 * @param {Date} currentDate - Current occurrence date
 * @param {Object} rrule - Parsed RRULE object
 * @param {Date} originalStartDate - Original event start date
 * @returns {Date} Next occurrence date
 */
function getNextOccurrence(currentDate, rrule, originalStartDate) {
  const next = new Date(currentDate)
  const freq = rrule.FREQ

  switch (freq) {
    case 'DAILY':
      const interval = rrule.INTERVAL ? parseInt(rrule.INTERVAL) : 1
      next.setDate(next.getDate() + interval)
      break

    case 'WEEKLY':
      const weekInterval = rrule.INTERVAL ? parseInt(rrule.INTERVAL) : 1
      if (rrule.BYDAY) {
        // Specific day(s) of week
        const days = rrule.BYDAY.split(',')
        const currentDay = next.getDay()
        const dayCodes = { 'SU': 0, 'MO': 1, 'TU': 2, 'WE': 3, 'TH': 4, 'FR': 5, 'SA': 6 }
        
        // Find next matching day
        let found = false
        for (const day of days) {
          const targetDay = dayCodes[day]
          if (targetDay > currentDay) {
            const daysToAdd = targetDay - currentDay
            next.setDate(next.getDate() + daysToAdd)
            found = true
            break
          }
        }
        
        if (!found) {
          // Move to next week
          const firstDay = dayCodes[days[0]]
          const daysToAdd = 7 - currentDay + firstDay + (weekInterval - 1) * 7
          next.setDate(next.getDate() + daysToAdd)
        }
      } else {
        // Same day of week, next interval
        next.setDate(next.getDate() + (7 * weekInterval))
      }
      break

    case 'MONTHLY':
      const monthInterval = rrule.INTERVAL ? parseInt(rrule.INTERVAL) : 1
      if (rrule.BYMONTHDAY) {
        // Specific day of month
        const dayOfMonth = parseInt(rrule.BYMONTHDAY)
        next.setMonth(next.getMonth() + monthInterval)
        next.setDate(dayOfMonth)
      } else {
        // Same day of month, next interval
        next.setMonth(next.getMonth() + monthInterval)
      }
      break

    case 'YEARLY':
      const yearInterval = rrule.INTERVAL ? parseInt(rrule.INTERVAL) : 1
      next.setFullYear(next.getFullYear() + yearInterval)
      break

    default:
      // Unknown frequency, default to daily
      next.setDate(next.getDate() + 1)
  }

  return next
}

/**
 * Expands all events (recurring and non-recurring) for a date range.
 * @param {Array} events - Array of events
 * @param {Date} rangeStart - Start of visible range
 * @param {Date} rangeEnd - End of visible range
 * @returns {Array} Expanded events array
 */
export function expandAllEvents(events, rangeStart, rangeEnd) {
  const expanded = []
  
  for (const event of events) {
    if (event.recurrenceRule) {
      const occurrences = expandRecurringEvent(event, rangeStart, rangeEnd)
      expanded.push(...occurrences)
    } else {
      expanded.push(event)
    }
  }
  
  return expanded
}

File: backend/src/main/java/com/agent/agenticcalendar/util/RecurrenceExpander.java
package com.agent.agenticcalendar.util;

import com.agent.agenticcalendar.model.Event;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;

/**
 * Utility class to expand recurring events into individual occurrences.
 * Implements Google Calendar-style recurrence handling using RFC 5545 RRULE format.
 * 
 * This class handles:
 * - FREQ (DAILY, WEEKLY, MONTHLY, YEARLY)
 * - INTERVAL (every N days/weeks/months/years)
 * - BYDAY (specific days of week)
 * - BYMONTHDAY (specific day of month)
 * - BYMONTH (specific months)
 * - UNTIL (end date)
 * - COUNT (number of occurrences)
 * - EXDATE (excluded dates - deleted instances)
 */
public class RecurrenceExpander {

    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE;
    private static final int MAX_OCCURRENCES = 1000; // Safety limit

    /**
     * Represents a parsed RRULE
     */
    public static class ParsedRRULE {
        public String FREQ;
        public Integer INTERVAL;
        public String BYDAY;
        public Integer BYMONTHDAY;
        public Integer BYMONTH;
        public LocalDate UNTIL;
        public Integer COUNT;
    }

    /**
     * Expands a recurring event into individual occurrences within a date range.
     * 
     * @param event The recurring event to expand
     * @param rangeStart Start of the date range
     * @param rangeEnd End of the date range
     * @return List of event occurrences (as Event objects with modified start/end times)
     */
    public static List<Event> expandEvent(Event event, LocalDateTime rangeStart, LocalDateTime rangeEnd) {
        if (event.getRecurrenceRule() == null || event.getRecurrenceRule().trim().isEmpty()) {
            // Not a recurring event - return as single event if in range
            if (isEventInRange(event, rangeStart, rangeEnd)) {
                return Collections.singletonList(event);
            }
            return Collections.emptyList();
        }

        ParsedRRULE rrule = parseRRULE(event.getRecurrenceRule());
        if (rrule == null) {
            // Invalid RRULE - return original event if in range
            if (isEventInRange(event, rangeStart, rangeEnd)) {
                return Collections.singletonList(event);
            }
            return Collections.emptyList();
        }

        Set<LocalDate> excludedDates = parseExcludedDates(event.getExcludedDates());
        LocalDateTime eventStart = event.getStartTime();
        LocalDateTime eventEnd = event.getEndTime();
        long durationMinutes = ChronoUnit.MINUTES.between(eventStart, eventEnd);

        List<Event> occurrences = new ArrayList<>();
        LocalDateTime currentOccurrence = eventStart;
        int occurrenceCount = 0;
        LocalDate rangeStartDate = rangeStart.toLocalDate();
        LocalDate rangeEndDate = rangeEnd.toLocalDate();

        // Start from the first occurrence that's >= rangeStart (or the original start)
        if (currentOccurrence.isBefore(rangeStart)) {
            // Fast-forward to first occurrence in range
            currentOccurrence = findNextOccurrence(eventStart, rrule, rangeStart.toLocalDate());
            if (currentOccurrence == null) {
                return Collections.emptyList();
            }
        }

        while (currentOccurrence.toLocalDate().isBefore(rangeEndDate.plusDays(1)) && 
               occurrenceCount < MAX_OCCURRENCES) {
            
            LocalDate occurrenceDate = currentOccurrence.toLocalDate();
            
            // Check if this occurrence is excluded
            if (!excludedDates.contains(occurrenceDate)) {
                // Check if occurrence is within range
                if (!occurrenceDate.isBefore(rangeStartDate) && !occurrenceDate.isAfter(rangeEndDate)) {
                    // Check UNTIL constraint
                    if (rrule.UNTIL != null && occurrenceDate.isAfter(rrule.UNTIL)) {
                        break;
                    }
                    
                    // Create occurrence
                    Event occurrence = createOccurrence(event, currentOccurrence, durationMinutes);
                    occurrences.add(occurrence);
                }
            }

            // Move to next occurrence
            currentOccurrence = getNextOccurrence(currentOccurrence, rrule, eventStart);
            if (currentOccurrence == null) {
                break;
            }
            
            occurrenceCount++;
            
            // Check COUNT constraint
            if (rrule.COUNT != null && occurrenceCount >= rrule.COUNT) {
                break;
            }
        }

        return occurrences;
    }

    /**
     * Checks if an event overlaps with a date range.
     */
    private static boolean isEventInRange(Event event, LocalDateTime rangeStart, LocalDateTime rangeEnd) {
        return !event.getEndTime().isBefore(rangeStart) && !event.getStartTime().isAfter(rangeEnd);
    }

    /**
     * Finds the first occurrence on or after the given date.
     */
    private static LocalDateTime findNextOccurrence(LocalDateTime eventStart, ParsedRRULE rrule, LocalDate targetDate) {
        LocalDateTime current = eventStart;
        int attempts = 0;
        
        while (current.toLocalDate().isBefore(targetDate) && attempts < MAX_OCCURRENCES) {
            current = getNextOccurrence(current, rrule, eventStart);
            if (current == null) {
                return null;
            }
            attempts++;
        }
        
        return current.toLocalDate().isBefore(targetDate) ? null : current;
    }

    /**
     * Creates an occurrence event from the master event.
     */
    private static Event createOccurrence(Event master, LocalDateTime occurrenceStart, long durationMinutes) {
        Event occurrence = new Event();
        occurrence.setId(master.getId());
        occurrence.setTitle(master.getTitle());
        occurrence.setDescription(master.getDescription());
        occurrence.setStartTime(occurrenceStart);
        occurrence.setEndTime(occurrenceStart.plusMinutes(durationMinutes));
        occurrence.setTimezone(master.getTimezone());
        occurrence.setType(master.getType());
        occurrence.setStatus(master.getStatus());
        occurrence.setLocation(master.getLocation());
        occurrence.setIsAllDay(master.getIsAllDay());
        occurrence.setRecurrenceRule(master.getRecurrenceRule());
        occurrence.setReminderTime(master.getReminderTime());
        occurrence.setColor(master.getColor());
        occurrence.setSessionId(master.getSessionId());
        occurrence.setGoogleUserId(master.getGoogleUserId());
        occurrence.setGoogleUserEmail(master.getGoogleUserEmail());
        return occurrence;
    }

    /**
     * Gets the next occurrence based on RRULE.
     */
    private static LocalDateTime getNextOccurrence(LocalDateTime current, ParsedRRULE rrule, LocalDateTime originalStart) {
        if (rrule.FREQ == null) {
            return null;
        }

        int interval = rrule.INTERVAL != null ? rrule.INTERVAL : 1;
        LocalDateTime next = current;

        switch (rrule.FREQ.toUpperCase()) {
            case "DAILY":
                next = current.plusDays(interval);
                break;

            case "WEEKLY":
                if (rrule.BYDAY != null && !rrule.BYDAY.isEmpty()) {
                    // Specific day(s) of week
                    List<DayOfWeek> targetDays = parseDaysOfWeek(rrule.BYDAY);
                    next = findNextDayOfWeek(current, targetDays, interval);
                } else {
                    // Same day of week, next interval
                    next = current.plusWeeks(interval);
                }
                break;

            case "MONTHLY":
                if (rrule.BYMONTHDAY != null) {
                    // Specific day of month
                    next = findNextMonthDay(current, rrule.BYMONTHDAY, interval);
                } else {
                    // Same day of month, next interval
                    next = current.plusMonths(interval);
                }
                break;

            case "YEARLY":
                if (rrule.BYMONTH != null && rrule.BYMONTHDAY != null) {
                    // Specific month and day
                    next = findNextYearDay(current, rrule.BYMONTH, rrule.BYMONTHDAY, interval);
                } else if (rrule.BYMONTH != null) {
                    // Specific month, same day
                    next = findNextYearMonth(current, rrule.BYMONTH, interval);
                } else {
                    // Same month and day, next interval
                    next = current.plusYears(interval);
                }
                break;

            default:
                return null;
        }

        return next;
    }

    /**
     * Finds the next occurrence matching specific days of week.
     */
    private static LocalDateTime findNextDayOfWeek(LocalDateTime current, List<DayOfWeek> targetDays, int interval) {
        LocalDateTime next = current;
        DayOfWeek currentDay = current.getDayOfWeek();
        
        // Find next matching day in current week
        for (DayOfWeek targetDay : targetDays) {
            if (targetDay.getValue() > currentDay.getValue()) {
                int daysToAdd = targetDay.getValue() - currentDay.getValue();
                return current.plusDays(daysToAdd);
            }
        }
        
        // Move to next week and find first matching day
        int daysToNextWeek = 7 - currentDay.getValue() + targetDays.get(0).getValue();
        next = current.plusDays(daysToNextWeek);
        
        // Apply interval
        if (interval > 1) {
            next = next.plusWeeks(interval - 1);
        }
        
        return next;
    }

    /**
     * Finds the next occurrence on a specific day of month.
     */
    private static LocalDateTime findNextMonthDay(LocalDateTime current, int targetDay, int interval) {
        LocalDateTime next = current.plusMonths(interval);
        int lastDayOfMonth = next.toLocalDate().lengthOfMonth();
        int dayToSet = Math.min(targetDay, lastDayOfMonth);
        return next.withDayOfMonth(dayToSet);
    }

    /**
     * Finds the next occurrence on a specific month and day.
     */
    private static LocalDateTime findNextYearDay(LocalDateTime current, int targetMonth, int targetDay, int interval) {
        LocalDateTime next = current.plusYears(interval);
        next = next.withMonth(targetMonth);
        int lastDayOfMonth = next.toLocalDate().lengthOfMonth();
        int dayToSet = Math.min(targetDay, lastDayOfMonth);
        return next.withDayOfMonth(dayToSet);
    }

    /**
     * Finds the next occurrence in a specific month.
     */
    private static LocalDateTime findNextYearMonth(LocalDateTime current, int targetMonth, int interval) {
        LocalDateTime next = current.plusYears(interval);
        return next.withMonth(targetMonth);
    }

    /**
     * Parses RRULE string into ParsedRRULE object.
     */
    public static ParsedRRULE parseRRULE(String rruleString) {
        if (rruleString == null || rruleString.trim().isEmpty()) {
            return null;
        }

        ParsedRRULE rrule = new ParsedRRULE();
        String[] parts = rruleString.toUpperCase().split(";");

        for (String part : parts) {
            String[] keyValue = part.split("=", 2);
            if (keyValue.length != 2) continue;

            String key = keyValue[0].trim();
            String value = keyValue[1].trim();

            switch (key) {
                case "FREQ":
                    rrule.FREQ = value;
                    break;
                case "INTERVAL":
                    try {
                        rrule.INTERVAL = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        // Invalid interval, ignore
                    }
                    break;
                case "BYDAY":
                    rrule.BYDAY = value;
                    break;
                case "BYMONTHDAY":
                    try {
                        rrule.BYMONTHDAY = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        // Invalid day, ignore
                    }
                    break;
                case "BYMONTH":
                    try {
                        rrule.BYMONTH = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        // Invalid month, ignore
                    }
                    break;
                case "UNTIL":
                    try {
                        // UNTIL can be in YYYYMMDD or YYYYMMDDTHHMMSS format
                        if (value.length() == 8) {
                            rrule.UNTIL = LocalDate.parse(value, DateTimeFormatter.BASIC_ISO_DATE);
                        } else if (value.length() >= 8) {
                            rrule.UNTIL = LocalDate.parse(value.substring(0, 8), DateTimeFormatter.BASIC_ISO_DATE);
                        }
                    } catch (Exception e) {
                        // Invalid date, ignore
                    }
                    break;
                case "COUNT":
                    try {
                        rrule.COUNT = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        // Invalid count, ignore
                    }
                    break;
            }
        }

        return rrule.FREQ != null ? rrule : null;
    }

    /**
     * Parses days of week from BYDAY value (e.g., "MO,TU,WE").
     */
    private static List<DayOfWeek> parseDaysOfWeek(String byday) {
        List<DayOfWeek> days = new ArrayList<>();
        String[] dayCodes = byday.split(",");
        
        Map<String, DayOfWeek> dayMap = new HashMap<>();
        dayMap.put("MO", DayOfWeek.MONDAY);
        dayMap.put("TU", DayOfWeek.TUESDAY);
        dayMap.put("WE", DayOfWeek.WEDNESDAY);
        dayMap.put("TH", DayOfWeek.THURSDAY);
        dayMap.put("FR", DayOfWeek.FRIDAY);
        dayMap.put("SA", DayOfWeek.SATURDAY);
        dayMap.put("SU", DayOfWeek.SUNDAY);
        
        for (String dayCode : dayCodes) {
            dayCode = dayCode.trim();
            // Handle BYSETPOS (e.g., "1MO" = first Monday)
            if (dayCode.length() > 2) {
                dayCode = dayCode.substring(dayCode.length() - 2);
            }
            DayOfWeek day = dayMap.get(dayCode);
            if (day != null) {
                days.add(day);
            }
        }
        
        return days;
    }

    /**
     * Parses excluded dates from comma-separated string.
     */
    private static Set<LocalDate> parseExcludedDates(String excludedDatesStr) {
        Set<LocalDate> excludedDates = new HashSet<>();
        if (excludedDatesStr == null || excludedDatesStr.trim().isEmpty()) {
            return excludedDates;
        }

        String[] dates = excludedDatesStr.split(",");
        for (String dateStr : dates) {
            try {
                LocalDate date = LocalDate.parse(dateStr.trim(), DATE_FORMATTER);
                excludedDates.add(date);
            } catch (Exception e) {
                // Invalid date format, skip
            }
        }

        return excludedDates;
    }

    /**
     * Checks if a recurring event has any occurrences in a date range.
     * This is used for efficient querying - we can filter recurring events
     * that might have occurrences in the range.
     */
    public static boolean hasOccurrenceInRange(Event event, LocalDateTime rangeStart, LocalDateTime rangeEnd) {
        if (event.getRecurrenceRule() == null) {
            return isEventInRange(event, rangeStart, rangeEnd);
        }

        ParsedRRULE rrule = parseRRULE(event.getRecurrenceRule());
        if (rrule == null) {
            return isEventInRange(event, rangeStart, rangeEnd);
        }

        // Check if event start is before range end (might have occurrences)
        if (event.getStartTime().isAfter(rangeEnd)) {
            return false;
        }

        // Check UNTIL constraint
        if (rrule.UNTIL != null && rrule.UNTIL.isBefore(rangeStart.toLocalDate())) {
            return false;
        }

        // For now, return true if event starts before range end
        // Full expansion will filter out actual occurrences
        return true;
    }
}

